samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java

```diff
@@ -191,13 +216,18 @@ private void updateOperatorGraphJson(MessageStreamImpl messageStream, Traversabl
    * @param opGraph {@link org.apache.samza.execution.JobGraphJsonGenerator.OperatorGraphJson}
    * @return {@link org.apache.samza.execution.JobGraphJsonGenerator.OperatorJson}
    */
-  private OperatorJson getOrCreateOperatorJson(OperatorSpec opSpec, OperatorGraphJson opGraph) {
+  private OperatorJson buildOperatorJson(OperatorSpec opSpec, OperatorGraphJson opGraph,
+      MessageStreamImpl messageStream) {
     Map<Integer, OperatorJson> operators = opGraph.operators;
     OperatorJson opJson = operators.get(opSpec.getOpId());
     if (opJson == null) {
       opJson = new OperatorJson();
       opJson.opCode = opSpec.getOpCode().name();
       opJson.opId = opSpec.getOpId();
+      StackTraceElement stackTraceElement = messageStream.getCallerStackTrace(opSpec.getOpId());
```

Comment:
Instead of the ExecutionPlanner needing to know the stacktrace, it'll be better if this is a field in the OperatorSpec directly. I.e., this class should just call opSpec.getSourceLine() or whatever this is called (not caller).

Even more generally, it'll be better to make this a part of the toString or toJson of each OperatorSpec. If we do that each OperatorSpec can provide additional information about itself, e.g. trigger information in windows or TTL for join.